### PR TITLE
refactors, onInvalidation hook, types.safeReference and more

### LINKED ISSUE #1088: types.safeReference/safeReferenceArray/safeReferenceMap
Given a pattern such as 
```
const Todo = ... with some id

const TodoStore = types.model({
  todos: types.array(Todo),
  selectedTodo: types.maybe(types.reference(Todo))
})
.actions(self => ({
  clearSelectedTodoRef() {
    self.selectedTodo = undefined
  },
  afterCreate() {
    // unselect a Todo when no longer available
    addDisposer(self, reaction(() => isValidReference(() => self.selectedTodo), (valid) => {
      if (!valid) this.clearSelectedTodo()
    });
  }
}))
```

Since that's such a common pattern it makes me wonder if there should be a types.safeReference(Todo) that automatically turns into undefined when the reference it is pointing to is no longer valid

That would solve "single row references"

For multiple rows, since with `types.array(types.safeReference(Todo))` array items would turn into undefined for those we would need something like a `types.safeReferenceArray(Todo)`, that would remove dead refs from the array automatically, and likewise `types.safeReferenceMap(Todo)`

### PR DESCRIPTION
**Pull request by xaviergonz**
Closes #1088 #1086 #1081

- small refactor over references
- small refactor over scalar and object nodes, by extracting common code to a BaseNode class
- added an internal mechanism to subscribe to hooks (all of them for object nodes, afterAttach/beforeDestroy for scalar nodes)
  - this should make it kind of trivial to make a `addHook(model/array/map/scalar/whateverType, 'afterCreation'..., (self) => {...})` in the future if desired
- added an onInvalidated option to the reference options (explained below)
- added a types.safeReference type (explained below)
- note that isValidReference and tryReference were alredy merged in #1087 this just adds extra docs for them

#### Reference validation: `isValidReference`, `tryReference`, `onInvalidate` hook and `types.safeReference`

Accessing an invalid reference (a reference to a dead/detached node) triggers an exception.

In order to check if a reference is valid, MST offers the `isValidReference(() => ref): boolean` function:
```ts
const isValid = isValidReference(() => store.myRef)
```

Also, if you are unsure if a reference is valid or not you can use the `tryReference(() => ref): ref | undefined` function:
```ts
// the result will be the passed ref if ok, or undefined if invalid
const maybeValidRef = tryReference(() => store.myRef)
```

The options parameter for references also accepts an optional `onInvalidated` hook, which will be called when the reference target node that the reference is pointing to is about to be detached/destroyed. It has the following signature:
```ts
const refWithOnInvalidated = types.reference(Todo, {
    onInvalidated(event: {
        // what is causing the target to become invalidated
        cause: "detach" | "destroy" | "invalidSnapshotIdentifier"
        // the target that is about to become invalidated (or undefined if the cause is "invalidSnapshotIdentifier")
        invalidTarget: STN | undefined
        // the identifier that is about to become invalidated
        invalidId: string | number
        // parent node of the reference (not the reference target)
        parent: IAnyStateTreeNode 
        // a function to remove the reference from its parent (or set to undefined in the case of models)
        removeRef: () => void 
        // a function to set our reference to a new target
        replaceRef: (newRef: STN | null | undefined) => void 
    }) {
        // do something
    }
})
```
Note that invalidation will only trigger while the reference is attached to a parent (be it a model, an array, a map, etc.).

A default implementation of such `onInvalidated` hook is provided by the `types.safeReference` type. It is like a standard reference, except that once the target node becomes invalidated it will:
- If its parent is a model: Set its own property to `undefined`
- If its parent is an array: Remove itself from the array
- If its parent is a map: Remove itself from the map

Strictly speaking it is implemented as
```js
types.maybe(types.reference(Type, {
    ...customGetSetIfAvailable,
    onInvalidated(ev) {
        ev.removeRef()
    }
}))
```

```js
const Todo = types.model({ id: types.identifier })
const Store = types.model({
    todos: types.array(Todo),
    selectedTodo: types.safeReference(Todo)
})

// given selectedTodo points to a valid Todo and that Todo is later removed from the todos
// array, then selectedTodo will automatically become undefined
```

### CODE REVIEW

**Comment by xaviergonz** [comment]
I was wondering, would `weakReference` be a better name than `safeReference`?

**Comment by xaviergonz** [comment]
all done

**Comment by xaviergonz** [comment]
all done

**Comment by k-g-a** [comment]
Hi, @xaviergonz! Great work, this feature is highly anticipated in our codebase. 
My only concern from the description is extracting a ```BaseNode``` - won't the performance decrease of introducing prototype chain for nodes be noticable? 
Will make a detailed review later today.

File: packages/mobx-state-tree/src/core/node/scalar-node.ts [hunk 1]
```diff
@@ -74,36 +63,54 @@ export class ScalarNode implements INode {
             const newPath = subpath === null ? "" : subpath
             if (this.subpath !== newPath) {
                 this.subpath = newPath
+                this.escapedSubpath = escapeJsonPath(this.subpath)
+                this.subpathAtom.reportChanged()
```
Comment (k-g-a) [inline packages/mobx-state-tree/src/core/node/scalar-node.ts]: Good catch! I wonder how did it work prior this? ) I guess reactions for splicing arrays of scalar nodes didn't fire? 
Should we add test for it?

File: packages/mobx-state-tree/src/core/node/scalar-node.ts [hunk 2]
```diff
@@ -74,36 +63,54 @@ export class ScalarNode implements INode {
...
             }
             if (newParent && newParent !== this.parent) {
+                // newParent.root.identifierCache!.mergeCache(this)
```
[the remark this answers is not in the document]
Comment (k-g-a) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: Is it possible to set a new parent for ```ScalarNode``` at all? As far as I remember there was a case to change ```subpath``` within the same ```parent``` or remove scalar from ```parent``` (```this.parent && !newParent``` case).

File: packages/mobx-state-tree/src/core/node/scalar-node.ts [hunk 3]
```diff
@@ -1,66 +1,55 @@
...
 
-    /*
-     * Returnes (escaped) path representation as string
-     */
-    public get path(): string {
-        if (!this.parent) return ""
-        return this.parent.path + "/" + escapeJsonPath(this.subpath)
-    }
+        this.state = NodeLifeCycle.CREATED
+        // there's no point in firing this event since it fires on the constructor, before
```
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: that comment is meant only for scalar-nodes, not for object-nodes, I'll change it to make it more clear

File: packages/mobx-state-tree/src/core/node/scalar-node.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: I think there actually isn't, should I change that to an exception?

File: packages/mobx-state-tree/src/core/node/BaseNode.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        return this.parent === null
+    }
+
+    abstract get root(): ObjectNode
+
+    abstract setParent(newParent: ObjectNode | null, subpath: string | null): void
+
+    protected abstract fireHook(name: Hook): void
+
+    protected internalFireHook(name: Hook) {
```
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: sounds good

[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: also renamed other "internalX" methods to "partialX" instead

File: packages/mobx-state-tree/src/core/node/Hook.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+/**
+ * @internal
+ * @private
+ */
+export enum Hook {
```
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/Hook.ts]: good catch, didn't see those

File: packages/mobx-state-tree/src/core/node/livelynessChecking.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+export type LivelynessMode = "warn" | "error" | "ignore"
```
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/livelynessChecking.ts]: will do

File: packages/mobx-state-tree/src/core/node/scalar-node.ts (hunk 1 above)
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: how to test this, making sure the scalar node ref didn't change after an applySnapshot?

**Comment by xaviergonz** [comment]
> My only concern from the description is extracting a BaseNode - won't the performance decrease of introducing prototype chain for nodes be noticable?

Do you think it can be that bad?

> I'd also add a test for the case where references array resides in separate tree. Something like:
> ```
> const Source = types.model({items: types.array(MyType)})
> const Refs = types.model({linkedItems: types.array(types.reference(MyType))})
> const source = Source.create({items: [{}, {}, {}]})
> const refs = Refs.create({linkedItems: source.items})
> ```

wouldn't that throw upon access? the refs and the source reside in a different "tree"

unless what you mean is that the array of weakRefs should become empty upon creation, but that'd mean that the weak refs would need to access the referenced nodes upon tree finalization, which would kind of defeat lazy loading?

right now it assumes that the snapshot is ok, should that be changed?

File: packages/mobx-state-tree/src/core/node/livelynessChecking.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/livelynessChecking.ts]: done

File: packages/mobx-state-tree/src/core/node/Hook.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/Hook.ts]: done

File: packages/mobx-state-tree/src/core/node/BaseNode.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/BaseNode.ts]: done

File: packages/mobx-state-tree/src/core/node/scalar-node.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: done, is that ok?

File: packages/mobx-state-tree/src/core/node/scalar-node.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: changed the comment

**Comment by k-g-a** [comment]
>Do you think it can be that bad?

I think that one level won't be noticable at all, but I am not sure, need measurement. 
Will try some profiling next week. But I think that this point should not block PR from being merged!

>wouldn't that throw upon access? the refs and the source reside in a different "tree"

I've just discovered, that this code was perfectly valid only up to 3.7.0 (including, https://codesandbox.io/s/vmq29xm1yl) - 3.7.1 broke it. Ability to have multiple trees is one of MST's top features. Reusing nodes from one tree in another was a bit of a pain, but it worked. 

>Found a way to do it without actually instantiating references (at least for the case without the custom get/set)

Didn't get that point. If one's reference source resides in another tree he must supply instances on create, not identifiers/snapshots. Othervise IdentifierCache should have been global and shared between all root nodes.
```ts
const Source = types.model({items: types.array(MyType)})
const Refs = types.model({linkedItems: types.array(types.reference(MyType))})
const source1 = Source.create({items: [{id: 1}, {id: 2}, {id: 3}]})
const source2 = Source.create({items: [{id: 1}, {id: 4}, {id: 5}]})
// at this point we could not supply ids, because id=1 lives in both source1 and source2
const refs = Refs.create({linkedItems: source1.items})
```

File: packages/mobx-state-tree/src/core/node/scalar-node.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (k-g-a) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: ok!

**Comment by xaviergonz** [comment]
> I think that one level won't be noticable at all, but I am not sure, need measurement.
Will try some profiling next week. But I think that this point should not block PR from being merged!

Fair enough :)

> I've just discovered, that this code was perfectly valid only up to 3.7.0 (including, https://codesandbox.io/s/vmq29xm1yl) - 3.7.1 broke it. Ability to have multiple trees is one of MST's top features. Reusing nodes from one tree in another was a bit of a pain, but it worked. Could we bring that behaviour back?

I didn't even know that was supposed to be possible. I guess that's something for @mweststrate to decide?

> Didn't get that point. If one's reference source resides in another tree he must supply instances on create, not identifiers/snapshots. Othervise IdentifierCache should have been global and shared between all root nodes.

Well, my answer was assuming that references and referenced nodes had to share a common parent. Of course if they don't there's no way check if references are valid at creation time.

**Comment by xaviergonz** [comment]
Hmm the docs say https://github.com/mobxjs/mobx-state-tree#references

> References are looked up through the entire tree but per type, so identifiers need to be unique in the entire tree.

**Comment by k-g-a** [comment]
>  taking your example code, what should this do?
Throw )) 
I know that using reference source from other tree is a kind of a hack and is only possible in limited scenarios. I won't even argue if this will be forbidden: there is always an alternative way through volatile state from env :)
If we really want to allow such approach, we could make shared IdentifierCache: this would be breaking change. Or won't if we make it under option or as a separate type. Anyway, I think that it's for @mweststrate to decide.

**Comment by mweststrate** [comment]
> Renamed `safeReference` to `weakReference`, it seems like a better name

Wouldn't this be confusing with the `weakReference` proposal in javascript? https://github.com/tc39/proposal-weakrefs/blob/master/specs/weakrefs.md

**Comment by mweststrate** [comment]
> > taking your example code, what should this do?
> 
> Throw ))
> I know that using reference source from other tree is a kind of a hack and is only possible in limited scenarios. I won't even argue if this will be forbidden: there is always an alternative way through volatile state from env :)
> If we really want to allow such approach, we could make shared IdentifierCache: this would be breaking change. Or won't if we make it under option or as a separate type. Anyway, I think that it's for @mweststrate to decide.

The idea I had was that standard `types.reference` doesn't work for different trees, however, that custom references would be able to resolve inside a different tree, by using resolveIdentifier and such:

```javascript
types.reference(User, {
        // given an identifier, find the user
        get(identifier) {
            return resolveIdentifier(someOtherStore /* this is another tree, accessed from closure, import, env or whatever */ , User /* type (living in the other tree) */, identifier)
        },
        // given a user, produce the identifier that should be stored
        set(value /* User */) {
            return value.name
        }
    })
```

**Comment by mweststrate** [comment]
@xaviergonz didn't review the code yet, since not all conceptual questions are answered :-). But this PR looks like a great improvement!

**Comment by xaviergonz** [comment]
@mweststrate Thanks for the feedback :)

> Wouldn't this be confusing with the weakReference proposal in javascript? 
Didn't even know that existed. I'll switch it back to safeReference then

> The idea I had was that standard types.reference doesn't work for different trees, however, that custom references would be able to resolve inside a different tree, by using resolveIdentifier and such:

Ah, so it is supposed to be ok for custom refs.
Ok, then I'll get rid of the "invalidSnapshotIdentifier" cause (since that needs snapshots of the reference targets on the same tree than references themselves) and add some tests for custom refs that point to other stores.

Thanks!

**Comment by xaviergonz** [comment]
Ok, I was able to keep the invalidSnapshotIdentifier even for custom identifiers in another stores
Also I wrote a unit test to make sure that custom identifiers can use nodes from external trees (it seems it keeps working)

**Comment by xaviergonz** [comment]
Also renamed it back to safeReference

**Comment by xaviergonz** [comment]
> ```
> const Source = types.model({items: types.array(MyType)})
> const Refs = types.model({linkedItems: types.array(types.reference(MyType))})
> const source = Source.create({items: [{}, {}, {}]})
> const refs = Refs.create({linkedItems: source.items})
> ```

@k-g-a added a unit test for that, but using custom references

File: packages/mobx-state-tree/src/core/node/scalar-node.ts (hunk 1 above)
Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: @k-g-a but getPath does not work over strings for example, I might be missing something...

Comment (xaviergonz) [reply packages/mobx-state-tree/src/core/node/scalar-node.ts]: could you write the unit test for this? 🙏

**Comment by xaviergonz** [comment]
I'll make a new release this afternoon

**Comment by xaviergonz** [comment]
@k-g-a 3.9.0 released

**Comment by k-g-a** [comment]
@xaviergonz, refactored our production code from volatile state to safeReferences: it works great and reduces boilerplate. Thanks for contribution!

**Comment by xaviergonz** [comment]
@k-g-a my pleasure :D

**Pull request merged**